### PR TITLE
Distributed insertion to one random shard

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add a new setting `insert_distributed_one_random_shard = 1` to allow insertion into multi-sharded distributed table without any distributed key.

Detailed description / Documentation draft:

EN document is updated. Others might need synchronization.

### CODE REVIEW

File: src/Storages/StorageDistributed.cpp [hunk 1]
```diff
@@ -890,6 +893,13 @@ void StorageDistributed::rename(const String & new_path_to_table_data, const Sto
 }
 
 
+size_t StorageDistributed::getRandomShardIndex(size_t shard_num)
+{
+    std::lock_guard lock(rng_mutex);
+    return std::uniform_int_distribution<size_t>(0, shard_num - 1)(rng);
```
Comment (alexey-milovidov) [inline src/Storages/StorageDistributed.cpp:899]: It does not support shard weights. I would like if it will work.

**Review by abyss7 — APPROVED** [review]

File: src/Storages/StorageDistributed.cpp [hunk 2]
```diff
@@ -893,10 +893,26 @@ void StorageDistributed::rename(const String & new_path_to_table_data, const Sto
 }
 
 
-size_t StorageDistributed::getRandomShardIndex(size_t shard_num)
+size_t StorageDistributed::getRandomShardIndex(const Cluster::ShardsInfo & shards)
 {
     std::lock_guard lock(rng_mutex);
```
Comment (amosbird) [inline src/Storages/StorageDistributed.cpp:898]: Ideally we can lock less)

Comment (abyss7) [reply src/Storages/StorageDistributed.cpp:898]: You're right - we need it only for `rng`

**Comment by abyss7** [comment]
Known problem with flaky mutation tests

**Pull request merged**